core/iwasm/aot/aot_loader.c

```diff
@@ -2764,6 +2765,97 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
     return ret;
 }
 
+#if WASM_ENABLE_LINUX_PERF != 0
+static uintptr_t
+get_func_addr(const AOTModule *module, void **sorted_func_ptrs, uint32 idx)
+{
+    return (uintptr_t)sorted_func_ptrs[idx];
+}
+
+static uint32
+get_func_size(const AOTModule *module, void **sorted_func_ptrs, uint32 idx)
+{
+    uint32 func_sz;
+
+    if (idx == module->func_count - 1)
+        func_sz = (uintptr_t)module->code + module->code_size
+                  - (uintptr_t)sorted_func_ptrs[idx];
+    else
+        func_sz = (uintptr_t)sorted_func_ptrs[idx + 1]
+                  - (uintptr_t)sorted_func_ptrs[idx];
+
+    return func_sz;
+}
+
+/* do a quick sort on func_ptrs in place*/
+static int
+compare_func_ptrs(const void *fp1, const void *fp2)
+{
+    return (intptr_t)(*(void **)fp1) - (intptr_t)(*(void **)fp2);
+}
+
+static void *
+sort_func_ptrs(const AOTModule *module, char *error_buf, uint32 error_buf_size)
+{
+    size_t content_len = sizeof(void *) * module->func_count;
+    void *sorted_func_ptrs =
+        loader_malloc(content_len, error_buf, error_buf_size);
+    if (!sorted_func_ptrs)
+        return NULL;
+
+    b_memcpy_s(sorted_func_ptrs, content_len, module->func_ptrs, content_len);
+
+    qsort(sorted_func_ptrs, module->func_count, sizeof(void *),
+          compare_func_ptrs);
+
+    return sorted_func_ptrs;
+}
+
+static bool
+create_perf_map(const AOTModule *module, char *error_buf, uint32 error_buf_size)
+{
+    void **sorted_func_ptrs = NULL;
+    char perf_map_info[128] = { 0 };
+    FILE *perf_map = NULL;
+    uint32 i;
+    pid_t pid = getpid();
+    bool ret = false;
+
+    sorted_func_ptrs = sort_func_ptrs(module, error_buf, error_buf_size);
+    if (!sorted_func_ptrs)
+        goto quit;
+
+    snprintf(perf_map_info, 128, "/tmp/perf-%d.map", pid);
+    perf_map = fopen(perf_map_info, "w");
+    if (!perf_map) {
+        LOG_WARNING("can't create /tmp/perf-%d.map, because %s", pid,
+                    strerror(errno));
+        goto quit;
+    }
+
+    for (i = 0; i < module->func_count; i++) {
+        memset(perf_map_info, 0, 128);
+        snprintf(perf_map_info, 128, "%lx  %x  aot_func#%u\n",
+                 get_func_addr(module, sorted_func_ptrs, i),
+                 get_func_size(module, sorted_func_ptrs, i), i);
```

Comment:
get_func_addr(.., i) returns sorted_func_ptrs[i], but it seems that we can not ensure that after sorted, `aot_func#i`'s func_ptr is sorted_func_ptrs[i].
Had better add an array to save the func indexes of each sorted func_ptr?